mirgecom/filter.py

```diff
@@ -23,24 +23,138 @@
 """
 import math
 import numpy as np
+import loopy as lp
+#from pytools import memoize_method
+from pytools.obj_array import obj_array_vectorized_n_args
 
 
-def get_spectral_filter(dim, order, cutoff, filter_order):
+def exponential_mode_response_function(mode_idx, alpha, cutoff, nfilt, filter_order):
+    return np.exp(-1.0 * alpha * ((sum(mode_idx) - cutoff) / nfilt)
+                  ** (2*filter_order))
+
+
+def make_spectral_filter(dim, order, cutoff, mode_response_function):
```

Comment:
Should take a `meshmode.discretization.ElementGroup` (or a subclass from `.poly_element`).

Reply:
I had trouble with this and the following related items due to mode_ids being attached to `master` for grudge/meshmode instead of pre-array-context.   I have some confusing modepy errors when I attempt to use grudge/meshmode master.  Check this out:
```
        from modepy import vandermonde
        for group in vol_discr.groups:
            mode_ids = group.mode_ids
>           vander = vandermonde(group.basis(), group.unit_nodes)

test_filter.py:119: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../meshmode/meshmode/discretization/poly_element.py:123: in basis
    mode_ids, basis = self._mode_ids_and_basis()
../../pytools/pytools/__init__.py:601: in wrapper
    result = function(obj, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <meshmode.discretization.poly_element.PolynomialWarpAndBlendElementGroup object at 0x145d3ffa0>

    @memoize_method
    def _mode_ids_and_basis(self):
        # for now, see https://gitlab.tiker.net/inducer/modepy/-/merge_requests/14
        import modepy.modes as modes
        if self.dim <= 3:
>           return modes.simplex_onb_with_mode_ids(self.dim, self.order)
E           AttributeError: module 'modepy.modes' has no attribute 'simplex_onb_with_mode_ids'

../../meshmode/meshmode/discretization/poly_element.py:118: AttributeError
============================================= short test summary info =============================================

```

Reply:
Yeah. This is not yet part of a modepy release, so you'll need to have git master installed.
```shell
conda uninstall modepy
pip uninstall modepy
git clone https://github.com/inducer/modepy.git
cd modepy
pip install -e .
```

Reply:
Ok, so using the mode_ids works terrifically. I checked in a form of this that uses it, but it also mixes up array-context dependent changes into the filter.   Currently, I am having trouble doing the following (in the context of array-context):
```
        input_mass = discr.zeros(actx) + 1.0
        input_energy = discr.zeros(actx) + 2.5
        input_momentum = [discr.zeros(actx) for i in range(dim)]
    
        uniform_soln = flat_obj_array(input_mass, input_energy, input_momentum)
>       filtered_soln = spectral_filter(discr=vol_discr, fields=uniform_soln)

test_filter.py:198: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
<decorator-gen-51>:2: in __call__
    ???
../../pytools/pytools/obj_array.py:208: in obj_array_vectorize_n_args
    result[i] = f(*new_args)
../mirgecom/filter.py:142: in __call__
    result = discr.empty(fields.array_context, dtype=fields.dtype)
../../meshmode/meshmode/discretization/__init__.py:279: in empty
    return self._new_array(actx, actx.empty, dtype=dtype)
../../meshmode/meshmode/discretization/__init__.py:264: in _new_array
    return _DOFArray.from_list(actx, [
../../meshmode/meshmode/discretization/__init__.py:265: in <listcomp>
    creation_func(shape=(grp.nelements, grp.nunit_dofs), dtype=dtype)
../../meshmode/meshmode/array_context.py:262: in empty
    return cla.empty(self.queue, shape=shape, dtype=dtype,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'Array' object has no attribute 'shape'") raised in repr()] Array object at 0x159aed100>
shape = (6, 35), dtype = dtype('O'), order = 'C', allocator = None, data = None, offset = 0, strides = (280, 8)
events = None

(yadda yadda yadda) 
    
        if _dtype_is_object(dtype):
>           raise TypeError("object arrays on the compute device are not allowed")
E           TypeError: object arrays on the compute device are not allowed

/Users/mtcampbe/miniforge3/envs/dgfem3/lib/python3.8/site-packages/pyopencl/array.py:469: TypeError
================================================ warnings summary =================================================

```